python/ray/data/_internal/execution/streaming_executor.py

```diff
@@ -219,10 +206,11 @@ def run(self):
                 pass
         except Exception as e:
             # Propagate it to the result iterator.
-            self._output_node.outqueue.append(e)
+            self._output_node.done_with_exception = e
```

Comment:
nit, let's add a method instead of manipulating the attributes. Could be something like `mark_finished(optional[Exception])`.
and internally we could use 2 attributes `_is_finished` and `_exception`. That would be make the code a bit easier to read.